pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java

```diff
@@ -127,34 +129,129 @@ public String getStatusDescription() {
+    }
+
+    protected List<String> getResourcesInCluster() {
+      return _helixAdmin.getResourcesInCluster(_clusterName);
+    }
+
+    protected ExternalView getResourceExternalView(String resourceName) {
+      return _helixAdmin.getResourceExternalView(_clusterName, resourceName);
+    }
+
+    protected CurrentState getCurrentState(String resourceName) {
+      PropertyKey.Builder keyBuilder = _helixDataAccessor.keyBuilder();
+      LiveInstance liveInstance = _helixDataAccessor.getProperty(keyBuilder.liveInstance(_instanceName));
+      String sessionId = liveInstance.getSessionId();
+      return _helixDataAccessor.getProperty(keyBuilder.currentState(_instanceName, sessionId, resourceName));
+    }
+  }
+
+  /**
+   * Service status callback that checks whether realtime consumption has caught up
+   * TODO: In this initial version, we are simply adding a configurable static wait time
+   * This can be made smarter:
+   * 1) Keep track of average consumption rate for table in server stats
+   * 2) Monitor consumption rate during startup, report GOOD when it stabilizes to average rate
+   * 3) Monitor consumption rate during startup, report GOOD if it is idle
+   */
+  public static class RealtimeConsumptionCatchupServiceStatusCallback extends BaseServiceStatusCallback {
+
+    private final Map<String, List<String>> _tableToConsumingSegmentsMap;
+    private long _endWaitTime = 0;
+
+    /**
+     * Realtime consumption catchup service which adds a static wait time for
+     */
+    public RealtimeConsumptionCatchupServiceStatusCallback(HelixManager helixManager, String clusterName, String instanceName,
+        long realtimeConsumptionCatchupWaitTimeSeconds) {
+      super(helixManager, clusterName, instanceName);
+
+      _tableToConsumingSegmentsMap = new HashMap<>();
+      for (String resourceName : getResourcesInCluster()) {
+        if (TableNameBuilder.isRealtimeTableResource(resourceName)) {
+
+          IdealState idealState = getResourceIdealState(resourceName);
+          if (idealState.isEnabled()) {
+            List<String> consumingSegments = new ArrayList<>();
+            for (String segmentName : idealState.getPartitionSet()) {
+              Map<String, String> instanceStateMap = idealState.getInstanceStateMap(segmentName);
+              if (CommonConstants.Helix.StateModel.RealtimeSegmentOnlineOfflineStateModel.CONSUMING.equals(
+                  instanceStateMap.get(instanceName))) {
+                consumingSegments.add(segmentName);
+              }
+            }
+            if (!consumingSegments.isEmpty()) {
+              _tableToConsumingSegmentsMap.put(resourceName, consumingSegments);
+            }
+          }
+        }
+      }
+
+      if (_tableToConsumingSegmentsMap.isEmpty()) {
+        LOGGER.info(
+            "No consuming segments to monitor on instance. Setting realtime consumption catchup wait time to 0ms");
+      } else {
+        long startTime = System.currentTimeMillis();
```

Comment:
Should you take the time when this method is entered?

It may also be worth commenting that even in the not-so-fancy determination of  the time a consuming segment is ready to serve queries, the computation should be:
 end-time = (time of creation of partition consumer) + (confugured max time to catch up).

We are sort of approximating it to :+1: 
end-time = (time of server startup) + (configured max timeto catch up)